# Introduction
The goal of this py is to answer the question from Refocus

**Tasks:**

1. **Process** data: remove complete duplicates, correct column names, etc;
2. **Calculate** net income as a separate column;
3. **Visualize** the dynamics of net profit in the South region.

**Answer format:** Ipynb file on GitHub with descriptions.

#### Loading Packages

In [13]:
conda update -n base -c defaults conda

Note: you may need to restart the kernel to use updated packages.
Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\Winata\anaconda3

  added / updated specs:
    - conda


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    aiohttp-3.8.1              |   py38h2bbff1b_1         495 KB
    aiosignal-1.2.0            |     pyhd3eb1b0_0          12 KB
    alabaster-0.7.12           |     pyhd3eb1b0_0          16 KB
    anaconda-client-1.11.0     |   py38haa95532_0         168 KB
    anaconda-project-0.11.1    |   py38haa95532_0         512 KB

    anyio-3.5.0                |   py38haa95532_0         166 KB
    argon2-cffi-21.3.0         |     pyhd3eb1b0_0          15 KB
    argon2-cffi-bindings-21.2.0|   py38h2bbff1b_0          36 KB
    asn1crypto-1.5.1           |   py38haa95532_0         167 KB
    astroid-2.11.7             |   py38haa95532_0    

  conda                               4.12.0-py38haa95532_0 --> 22.9.0-py38haa95532_0
  conda-content-tru~ pkgs/main/noarch::conda-content-trust~ --> pkgs/main/win-64::conda-content-trust-0.1.3-py38haa95532_0
  conda-package-han~                   1.7.2-py38h76e460a_0 --> 1.9.0-py38h8cc25b3_1
  conda-repo-cli     pkgs/main/noarch::conda-repo-cli-1.0.~ --> pkgs/main/win-64::conda-repo-cli-1.0.24-py38haa95532_0
  conda-token                            0.3.0-pyhd3eb1b0_0 --> 0.4.0-pyhd3eb1b0_0
  contextlib2        pkgs/main/noarch::contextlib2-0.6.0.p~ --> pkgs/main/win-64::contextlib2-21.6.0-py38haa95532_0
  cryptography                         3.4.8-py38h71e12ea_0 --> 38.0.1-py38h21b164f_0
  curl                                    7.71.1-h2a8f88b_1 --> 7.85.0-h2bbff1b_0
  cycler             pkgs/main/win-64::cycler-0.10.0-py38_0 --> pkgs/main/noarch::cycler-0.11.0-pyhd3eb1b0_0
  cython             conda-forge::cython-0.29.24-py38h885f~ --> pkgs/main::cython-0.29.32-py38hd77b12b_0
  cyto

In [6]:
conda install -c conda-forge matplotlib

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\Winata\anaconda3

  added / updated specs:
    - matplotlib


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2022.9.24  |       h5b45459_0         189 KB  conda-forge
    certifi-2022.9.24          |     pyhd8ed1ab_0         155 KB  conda-forge
    charset-normalizer-2.0.4   |     pyhd3eb1b0_0          35 KB
    google-api-core-2.2.2      |     pyhd3eb1b0_0          69 KB
    google-api-core-grpc-2.2.2 |       hd3eb1b0_0          14 KB
    google-cloud-core-2.2.2    |     pyhd3eb1b0_0          26 KB
    googleapis-common-protos-1.57.0|   py38h5b57dd5_0         130 KB  conda-forge
    grpcio-1.42.0              |   py38hc60d5dd_0         1.9 MB
    grpcio-status-1.41.1       |     pyhd3eb1b0_0          11 KB
    idna-3.4                   |     pyhd8ed1ab_0          55 KB  conda-


The environment is inconsistent, please check the package plan carefully
The following packages are causing the inconsistency:

  - conda-forge/win-64::aiohttp==3.8.1=py38h294d835_0
  - defaults/win-64::anaconda-client==1.7.2=py38_0
  - defaults/win-64::anaconda==custom=py38_1
  - defaults/win-64::anaconda-navigator==2.1.4=py38haa95532_0
  - defaults/noarch::anaconda-project==0.8.4=py_0
  - defaults/win-64::argon2-cffi==20.1.0=py38he774522_1
  - defaults/win-64::arrow-cpp==3.0.0=py38h3dd2304_4
  - defaults/win-64::astroid==2.4.2=py38_0
  - defaults/win-64::astropy==4.0.2=py38he774522_0
  - defaults/win-64::bcrypt==3.2.0=py38he774522_0
  - defaults/win-64::bkcharts==0.2=py38_0



certifi-2022.9.24    | 155 KB    |            |   0% 
certifi-2022.9.24    | 155 KB    | ########## | 100% 
certifi-2022.9.24    | 155 KB    | ########## | 100% 

patsy-0.5.3          | 189 KB    |            |   0% 
patsy-0.5.3          | 189 KB    | ########## | 100% 
patsy-0.5.3          | 189 KB    | ########## | 100% 
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done


  - defaults/noarch::bleach==3.2.1=py_0
  - defaults/win-64::bokeh==2.2.3=py38_0
  - defaults/win-64::bottleneck==1.3.2=py38h2a96729_1
  - defaults/win-64::conda==4.12.0=py38haa95532_0
  - defaults/win-64::conda-build==3.20.5=py38_1
  - defaults/noarch::conda-content-trust==0.1.1=pyhd3eb1b0_0
  - defaults/win-64::conda-package-handling==1.7.2=py38h76e460a_0
  - defaults/noarch::conda-repo-cli==1.0.4=pyhd3eb1b0_0
  - defaults/noarch::conda-token==0.3.0=pyhd3eb1b0_0
  - defaults/noarch::conda-verify==3.4.2=py_1
  - defaults/win-64::cycler==0.10.0=py38_0
  - defaults/noarch::dask==2.30.0=py_0
  - conda-forge/noarch::google-api-core-grpc==1.31.2=hd8ed1ab_1
  - defaults/noarch::google-auth==1.28.0=pyhd3eb1b0_0
  - conda-forge/noarch::google-cloud-bigquery==2.24.1=pyhd3deb0d_0
  - conda-forge/noarch::google-cloud-bigquery-core==2.24.1=pyhd3deb0d_0
  - conda-forge/noarch::google-cloud-bigquery-storage-core==2.10.0=pyh6c4a22f_0
  - conda-forge/noarch::google-cloud-core==1.7.2=pyh6c4a22f_0
  - 

In [11]:
pip install -U basemap

     -------------------------------------- 488.0/488.0 kB 2.5 MB/s eta 0:00:00
     ---------------------------------------- 30.5/30.5 MB 3.6 MB/s eta 0:00:00
     ---------------------------------------- 14.7/14.7 MB 3.6 MB/s eta 0:00:00
     -------------------------------------- 46.5/46.5 kB 789.1 kB/s eta 0:00:00
     ---------------------------------------- 4.8/4.8 MB 3.5 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 1.19.2
    Uninstalling numpy-1.19.2:
      Successfully uninstalled numpy-1.19.2
Note: you may need to restart the kernel to use updated packages.


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gensim 4.1.2 requires Cython==0.29.23, but you have cython 0.29.24 which is incompatible.

[notice] A new release of pip available: 22.2.1 -> 22.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [14]:
import pandas as pd # data processing
import numpy as np # linear algebra
from datetime import datetime
import matplotlib.pyplot as plt # used for graphical representation 
import seaborn as sns # used for graphical representation, seaborn is build on the top of mat

ImportError: cannot import name 'dedent' from 'matplotlib.cbook' (C:\Users\Winata\anaconda3\lib\site-packages\matplotlib\cbook\__init__.py)

#### Loading Data

In [498]:
df = pd.read_csv("python - python.csv")

#### First Checking

In [499]:
df.head(5)

Order ID  Order Date  Ship Date       Ship Mode Customer ID  \
0  CA-2017-152156     43047.0    43050.0    Second Class    CG-12520   
1  CA-2017-152156     43047.0    43050.0    Second Class    CG-12520   
2  CA-2017-138688     42898.0    42902.0    Second Class    DV-13045   
3  US-2016-108966     42654.0    42661.0  Standard Class    SO-20335   
4  US-2016-108966     42654.0    42661.0  Standard Class    SO-20335   

     Customer Name    Segment        Country             City       State  \
0      Claire Gute   Consumer  United States        Henderson    Kentucky   
1      Claire Gute   Consumer  United States        Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States      Los Angeles  California   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   

   ...                                              X9283         Sales  \
0  ...                  Bush Somerset Collection Bookcase        261.96   
1  ...      Hon Deluxe Fabric Upholstered Stacking Chairs  Rounded Back   
2  ...  Self-Adhesive Address Labels for Typewriters b...         14.62   
3  ...      Bretford CR4500 Series Slim Rectangular Table      957.5775   
4  ...                     Eldon Fold 'N Roll Cart System        22.368   

  Quantity Discount Returned Unnamed: 20 Unnamed: 21 Unnamed: 22 Unnamed: 23  \
0        2        0        0         NaN         NaN         NaN         NaN   
1   731.94        3        0           0         NaN         NaN         NaN   
2        2        0        0         NaN         NaN         NaN         NaN   
3        5     0.45        0         NaN         NaN         NaN         NaN   
4        2      0.2        0         NaN         NaN         NaN         NaN   

  Unnamed: 24  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  

[5 rows x 25 columns]

In [500]:
df.shape

(10797, 25)

we have 25 variables and 10,797 rows. Let's see if there are duplicate rows in the training data

In [501]:
df = df.drop_duplicates()

From our first look, we see there is duplicates. Let's take a look to others info

In [502]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9993 entries, 0 to 9993
Data columns (total 25 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Order ID       9993 non-null   object 
 1   Order Date     9176 non-null   float64
 2   Ship Date      9176 non-null   float64
 3   Ship Mode      9176 non-null   object 
 4   Customer ID    9176 non-null   object 
 5   Customer Name  9176 non-null   object 
 6   Segment        9176 non-null   object 
 7   Country        9176 non-null   object 
 8   City           9176 non-null   object 
 9   State          9176 non-null   object 
 10  Postal Code    9165 non-null   float64
 11  Region         9176 non-null   object 
 12  Product ID     9176 non-null   object 
 13  X9281          9176 non-null   object 
 14  X9282          9176 non-null   object 
 15  X9283          9176 non-null   object 
 16  Sales          9176 non-null   object 
 17  Quantity       9169 non-null   object 
 18  Discount

From info() method we see that the data type is float or object. No null values are present in the data set.
We need to see that the data only from Order ID until Returned.
So we will drop other rows "Unnamed: .." and change the data to the right format.

In [503]:
ndf = df.iloc[:,0:20].drop_duplicates()

In [504]:
ndf.shape

(9993, 20)

Now let's change the format of our data

- > Date | Order Date, Ship Date
- > Int | Postal Code
- > Float | Sales, Discount, Quantity, Returned


Name:

We will also change the name of this colomns, so it will help user understand our data

- X9281 - Product Category
- X9282 - Product
- X9283 - Product Detail

Date Format

In [505]:
ndf["Order Date"] = pd.to_datetime(ndf['Order Date'], unit='D', origin='1899-12-30')
ndf["Ship Date"] = pd.to_datetime(ndf['Ship Date'], unit='D', origin='1899-12-30')

Int format

In [506]:
ndf['Postal Code'] = pd.to_numeric(ndf['Postal Code'], errors = 'coerce').fillna(0).astype(int)

Float Format

In [507]:
ndf[['Sales', 'Quantity', 'Discount', 'Returned']] = ndf[['Sales', 'Quantity', 'Discount', 'Returned']].apply(pd.to_numeric, errors = 'coerce') 

Change the name of our columns

In [508]:
ndf.rename(columns = {'X9281' : 'Product Category', 'X9282' : 'Product', 'X9283' : 'Product Detail'}, inplace = True)

More Deeper Cleansing

In [509]:
unclean_df = ndf[ndf['Order ID'].str.contains(',')]
unclean_df.shape

(817, 20)

From the data we can see there is 817 data, that we fail to read.
When we see the pattern, we can know the problem is because of all data in first column.
That's why we need to do some cleansing to this data

In [510]:
unclean_df.head(2)

Order ID Order Date Ship Date  \
16  CA-2015-105893,42319,42326,Standard Class,PK-1...        NaT       NaT   
20  CA-2015-143336,42243,42248,Second Class,ZD-219...        NaT       NaT   

   Ship Mode Customer ID Customer Name Segment Country City State  \
16       NaN         NaN           NaN     NaN     NaN  NaN   NaN   
20       NaN         NaN           NaN     NaN     NaN  NaN   NaN   

    Postal Code Region Product ID Product Category Product Product Detail  \
16            0    NaN        NaN              NaN     NaN            NaN   
20            0    NaN        NaN              NaN     NaN            NaN   

    Sales  Quantity  Discount  Returned  
16    NaN       NaN       NaN       NaN  
20    NaN       NaN       NaN       NaN

In [511]:
# We spot the pattern
# Since we using "," for split data, we suspect that Product Detail will have "," too. So we will split it by 2
unclean_df_start = pd.DataFrame(unclean_df["Order ID"].str.split(",", n = 15, expand = True)) # Fist data before Product Detail
unclean_df_last = unclean_df_start.iloc[:,-1].str.rsplit(",",4, expand = True) # Rest of data
unclean_ndf = pd.concat([unclean_df_start.iloc[:,:-1], unclean_df_last], axis=1) # Joining first and last data
unclean_ndf.columns = ndf.columns[:] # Change the columns names

We need to clean our data **Unclean_ndf**

In [512]:
unclean_ndf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 817 entries, 16 to 9992
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          817 non-null    object
 1   Order Date        817 non-null    object
 2   Ship Date         817 non-null    object
 3   Ship Mode         817 non-null    object
 4   Customer ID       817 non-null    object
 5   Customer Name     817 non-null    object
 6   Segment           817 non-null    object
 7   Country           817 non-null    object
 8   City              817 non-null    object
 9   State             817 non-null    object
 10  Postal Code       817 non-null    object
 11  Region            817 non-null    object
 12  Product ID        817 non-null    object
 13  Product Category  817 non-null    object
 14  Product           817 non-null    object
 15  Product Detail    817 non-null    object
 16  Sales             817 non-null    object
 17  Quantity      

In [513]:
# Change the date to int format
unclean_ndf['Order Date'] = pd.to_numeric(unclean_ndf['Order Date'], errors = 'coerce').fillna(0).astype(int)
unclean_ndf['Ship Date'] = pd.to_numeric(unclean_ndf['Ship Date'], errors = 'coerce').fillna(0).astype(int)
# Change the int format to date
unclean_ndf["Order Date"] = pd.to_datetime(unclean_ndf['Order Date'], unit='D', origin='1899-12-30')
unclean_ndf["Ship Date"] = pd.to_datetime(unclean_ndf['Ship Date'], unit='D', origin='1899-12-30')

# Postal Code
unclean_ndf['Postal Code'] = pd.to_numeric(unclean_ndf['Postal Code'], errors = 'coerce').fillna(0).astype(int)

# Change to Float format
unclean_ndf[['Sales', 'Quantity', 'Discount', 'Returned']] = unclean_ndf[['Sales', 'Quantity', 'Discount', 'Returned']].apply(pd.to_numeric, errors = 'coerce') 

In [518]:
clean_df = pd.concat([ndf[~ndf['Order ID'].str.contains(',')], unclean_ndf]) # Joining our data

In this step we already doing the hard part (Cleansing the data). We will make sure to check one more time the columns and rows number that we have

In [521]:
print(clean_df.shape, ndf.shape)

(9993, 20) (9993, 20)


Last, let's drop column that have NA in Sales columns

In [533]:
clean_df = clean_df.dropna(subset=['Sales'])

### Congratulation! Our first task is done!
Now let's go to the next task

2. **Calculate** net income as a separate column;
3. **Visualize** the dynamics of net profit in the South region.

Need to create new column for net income
we will using 
> **((Sales x Quantity) - (Sales x Quantity x discount)) - ((Sales x Returned) - (Sales x Returned) x discount)**

2. Calculate net income as a separate column;

In [556]:
# Get Net Income
clean_df["Net Income"] = (clean_df["Sales"]*clean_df["Quantity"]) - (clean_df["Sales"]*clean_df["Quantity"]*clean_df["Discount"]) - (clean_df["Sales"]*clean_df["Returned"]) - (clean_df["Sales"]*clean_df["Returned"]*clean_df["Discount"])

<ipython-input-556-321865ac9764>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_df["Net Income"] = (clean_df["Sales"]*clean_df["Quantity"]) - (clean_df["Sales"]*clean_df["Quantity"]*clean_df["Discount"]) - (clean_df["Sales"]*clean_df["Returned"]) - (clean_df["Sales"]*clean_df["Returned"]*clean_df["Discount"])


In [560]:
clean_df["Net Income"].head()

0     523.920000
2      29.240000
3    2633.338125
4      35.788800
6      29.120000
Name: Net Income, dtype: float64

3. Visualize the dynamics of net profit in the South region.

In [566]:
clean_df.shape

(8302, 21)

In [568]:
south_df = clean_df[clean_df["Region"] == "South"]

In [571]:
sns.distplot(south_df['Net Income'])

NameError: name 'sns' is not defined